This project will serve to analyze the data given to find out some aspects about the people in it, we will look into seeing what percent of the about 1000 people have children, are smokers, where do they mostly live and the relations of that into their insurance cost

In [22]:
import csv

#create all the values from the headers as lists
ages = []
gender = []
bmi = []
children = []
smoker = []
region = []
charges = []

#first we will look to find out the average age of people in the study
with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    for row in insurance_reader:
        ages.append(int(row['age']))

average_age = sum(ages) / len(ages)

print('The average age is ' + str(round(average_age, 1)) + ' years old')
    

The average age is 39.2 years old


In [39]:
#find out the gender distribution

with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    male_counter = 0
    female_counter = 0
    for row in insurance_reader:
        gender.append(row['sex'])
        if row['sex'] == 'male':
            male_counter += 1
        else:
            female_counter += 1
            
#calculate the percentage of each          
male_percentage = (male_counter / (male_counter+female_counter)) * 100
female_percentage = 100 - male_percentage
            
print("There are about " + str(round(male_percentage, 2)) + '%' + " of man in this study and " + str(round(female_percentage, 2)) + '% of woman') 

There are about 50.52% of man in this study and 49.48% of woman


In [42]:
#now we will look into the average amount of children, using pretty much the same code as age

with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    for row in insurance_reader:
        children.append(int(row['children']))

average_children = sum(children) / len(children)

print('The average amount of children is ' + str(round(average_children, 1)))

The average amount of children is 1.1


In [46]:
#find out how many are smokers, again same code as male/female

with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    smoker_counter = 0
    nonsmoker_counter = 0
    for row in insurance_reader:
        smoker.append(row['smoker'])
        if row['smoker'] == 'yes':
            smoker_counter += 1
        else:
            nonsmoker_counter += 1
            
#calculate the percentage of each          
smoker_percentage = (smoker_counter / (smoker_counter+nonsmoker_counter)) * 100
nonsmoker_percentage = 100 - smoker_percentage
            
print("There are about " + str(round(smoker_percentage, 2)) + '%' + " of smokers in this study and " + str(round(nonsmoker_percentage, 2)) + '% of non-smokers')



There are about 20.48% of smokers in this study and 79.52% of non-smokers


In [52]:
#find out the average BMI 

with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    for row in insurance_reader:
        bmi.append(float(row['bmi']))

average_bmi = sum(bmi) / len(bmi)

print('The average bmi is ' + str(round(average_bmi,2)))

The average bmi is 30.66


In [83]:
#find the most common region people live in:
region_counter = {}

with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    for row in insurance_reader:
        region_name = row['region']
        region.append(row['region'])
        if region_name not in region_counter:
            region_counter[region_name] = 1
        else:         
            region_counter[region_name] += 1
        

#Show the results
print("Regions and their respective people counts:")
for region_name, count in region_counter.items():
    print(f"{region_name}: {count} people")

Regions and their respective people counts:
southwest: 325 people
southeast: 364 people
northwest: 325 people
northeast: 324 people


In [88]:
#now to analyze the charges and find the average
with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    for row in insurance_reader:
        charges.append(float(row['charges']))

average_charges = sum(charges) / len(charges)

print('The average charge is ' + str(round(average_charges,2)) + ' dollars')

The average charge is 13270.42 dollars


After this basic analysis of the data we can see that:
The average patient in this study is 39 year old non-smoking man with 1 children who lives in the southeast area and pays asbout 13270.42 dollars for his insurance.

Now we will get into the really fun part, going into the meat of this data, understand the difference in charges for man and woman, smoking and non-smoking, what impact do children and BMI have in the cost, let's go.


In [105]:
#starting with the difference cost of man and woman
male_costs = []
female_costs = []
with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    for row in insurance_reader:
        if row['sex'] == 'male':
            male_costs.append(float(row['charges']))
        else:
            female_costs.append(float(row['charges']))

average_charges_male = sum(male_costs) / len(male_costs)
average_charges_female = sum(female_costs) / len(female_costs)

print('A man pays on average ' + str(round(average_charges_male,2)) + '$ for his insurance')
print('A woman pays on average ' + str(round(average_charges_female,2)) + '$ for her insurance')



A man pays on average 13956.75$ for his insurance
A woman pays on average 12569.58$ for her insurance


In [119]:
#difference in smoking and non-smoking plus smoking and non-smoking mand and woman
smoking_charges = []
nonsmoking_charges = []
male_smoking_charges = []
female_smoking_charges = []
male_nonsmoking_charges = []
female_nonsmoking_charges = []



with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    for row in insurance_reader:
        if row['smoker'] == 'yes':
            smoking_charges.append(float(row['charges']))
        else:
            nonsmoking_charges.append(float(row['charges']))
            
        if row['smoker'] == 'yes' and row['sex'] == 'male':
            male_smoking_charges.append(float(row['charges']))
        elif row['smoker'] == 'yes' and row['sex'] == 'female':
            female_smoking_charges.append(float(row['charges']))
        elif row['smoker'] == 'no' and row['sex'] == 'male':
            male_nonsmoking_charges.append(float(row['charges']))
        elif row['smoker'] == 'no' and row['sex'] == 'female':
            female_nonsmoking_charges.append(float(row['charges']))
            
def find_average(list):
    average = sum(list) / len(list)
    return round(average, 2)

print("The average insurance cost for a smoker is " + str(find_average(smoking_charges)) + ' dollars')
print("The average insurance cost for a nonsmoker is " + str(find_average(nonsmoking_charges)) + ' dollars')
print("The average insurance cost for a smoking male is " + str(find_average(male_smoking_charges)) + ' dollars')
print("The average insurance cost for a smoking female is " + str(find_average(female_smoking_charges)) + ' dollars')
print("The average insurance cost for a nonsmoking male is " + str(find_average(male_nonsmoking_charges)) + ' dollars')
print("The average insurance cost for a nonsmoking female is " + str(find_average(female_nonsmoking_charges)) + ' dollars')

The average insurance cost for a smoker is 32050.23 dollars
The average insurance cost for a nonsmoker is 8434.27 dollars
The average insurance cost for a smoking male is 33042.01 dollars
The average insurance cost for a smoking female is 30679.0 dollars
The average insurance cost for a nonsmoking male is 8087.2 dollars
The average insurance cost for a nonsmoking female is 8762.3 dollars


In [120]:
#finding out the same for children and its amount and gender

charges_by_children = {}

with open('insurance.csv') as insurance_file:
    insurance_reader = csv.DictReader(insurance_file)
    for row in insurance_reader:
        num_children = int(row['children'])
        charges = float(row['charges'])
        
        # Check if the number of children is already a key in the dictionary
        if num_children in charges_by_children:
            charges_by_children[num_children].append(charges)
        else:
            # If not, create a new key with a list containing the charge
            charges_by_children[num_children] = [charges]

# Calculate the average charges for each number of children
average_charges_by_children = {}
for num_children, charge_list in charges_by_children.items():
    average_charge = sum(charge_list) / len(charge_list)
    average_charges_by_children[num_children] = average_charge

# Print the results
for num_children, average_charge in average_charges_by_children.items():
    print(f'Average charges for {num_children} children: {round(average_charge, 2)} dollars')

Average charges for 0 children: 12365.98 dollars
Average charges for 1 children: 12731.17 dollars
Average charges for 3 children: 15355.32 dollars
Average charges for 2 children: 15073.56 dollars
Average charges for 5 children: 8786.04 dollars
Average charges for 4 children: 13850.66 dollars


In [122]:
#now to finish this off I made a function that allows you to look for a specific type of person and return the average cost for them

def calculate_average_insurance_cost(gender=None, num_children=None, smoker=None, bmi=None, region=None):
    charges = []
    
    with open('insurance.csv') as insurance_file:
        insurance_reader = csv.DictReader(insurance_file)
        for row in insurance_reader:
            match_criteria = True
            
            if gender is not None and row['sex'] != gender:
                match_criteria = False
            if num_children is not None and int(row['children']) != num_children:
                match_criteria = False
            if smoker is not None and row['smoker'] != smoker:
                match_criteria = False
            if bmi is not None and float(row['bmi']) != bmi:
                match_criteria = False
            if region is not None and row['region'] != region:
                match_criteria = False
                
            if match_criteria:
                charges.append(float(row['charges']))
    
    if charges:
        average_charge = sum(charges) / len(charges)
        return round(average_charge, 2)
    else:
        return 'No data found for the specified criteria.'

# Example usage without specifying BMI:
average_cost = calculate_average_insurance_cost('male', 1, 'yes', region='southeast')
print(f'Average insurance cost: ${average_cost}')

average_cost = calculate_average_insurance_cost('female', 0, 'no')
print(f'Average insurance cost: ${average_cost}')


Average insurance cost: $35925.52
Average insurance cost: $7688.32
